src/main/java/sk/freemap/gpxAnimator/TileCache.java

```diff
@@ -0,0 +1,135 @@
+ *  2. If no cache specified, then no caching occurs.
+ *  3. Cache design assumes relatively few files (within OS limits of
+ *     number of files per directory).
+ *  4. Cache file names are based on a hash of the URL. This should remove
+ *     any naming issues because of special characters in the URL that might
+ *     be incompatible with the OS file naming restrictions. It also allows
+ *     for the cache to hold files for more than one TMS or for more than
+ *     one zoom level, etc.
+ *  5. Cache files age out based on a user specified age limit. By default
+ *     this is quite large (12 hours) because with the possible exception of
+ *     tiles from the default renderer at openstreetmap.org map tile generation
+ *     is not done very often (might be a month or two between changes).
+ */
+
+package sk.freemap.gpxAnimator;
+
+import java.awt.image.BufferedImage;
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Date;
+import javax.imageio.ImageIO;
+import org.apache.commons.io.FileUtils;
+
+public class TileCache {
+
+    public static BufferedImage getTile( String url, String cachePath, Long cacheTimeLimit ) throws UserException {
+
+        if (cachePath == null) {
+            //System.out.println("No cache path defined, downloading " + url);
+            return unCachedGetTile( url );
+        }
+
+        //System.out.println("Caching enabled " + url);
+        return cachedGetTile( url, cachePath, cacheTimeLimit );
+    }
+
+    private static BufferedImage unCachedGetTile( String url ) throws UserException {
+        BufferedImage mapTile;
+        
+        System.setProperty("http.agent", "GPX Animator " + Constants.VERSION);
+        try {
+            mapTile = ImageIO.read(new URL(url));
+        } catch (final IOException e) {
+            throw new UserException("error reading tile " + url, e);
+        }
+        if (mapTile == null) {
+            throw new UserException("could not read tile " + url);
+        }
+        
+        return mapTile;
+    }
+    
+    private static BufferedImage cachedGetTile( String url, String cachePath, Long cacheTimeLimit ) throws UserException {
+        BufferedImage mapTile;
+        String fname = hashName(url);
+        String path = cachePath + File.separator + fname;
+        File f = new File(path);
+        
+        // Age out old file if needed
+        if (f.isFile()){
+            Date fileDate = new Date(f.lastModified());
+            long msBetweenDates = new Date().getTime() - fileDate.getTime();
+            if ((msBetweenDates/1000) > cacheTimeLimit) {
+                System.out.println("Removing cache entry for " + url); 
+                if (!f.delete()) {
+                    throw new UserException("failed to delete cache entry for " + url);
```

Comment:
Maybe it is a good idea to add the name of the cache file to the error message because it is different from the URL. With the file name, the user could then check the cache directory and the file for problems.